packages/web/components/input/limit-input.tsx

```diff
@@ -0,0 +1,448 @@
+    setAmountSafe,
+  ]);
+
+  useEffect(() => {
+    if (focused !== FocusedInput.FIAT || !price) return;
+
+    const value = fiatAmount && fiatAmount.length > 0 ? fiatAmount : undefined;
+    const tokenValue = value ? new Dec(value).quo(price) : undefined;
+    setAmountSafe("token", tokenValue ? tokenValue.toString() : undefined);
+  }, [price, fiatAmount, setAmountSafe, focused]);
+
+  return (
+    <div className="relative h-[108px]">
+      {(["fiat", "token"] as ("fiat" | "token")[]).map((inputType) => (
+        <AutoInput
+          key={inputType}
+          type={inputType}
+          baseAsset={baseAsset}
+          focused={focused}
+          swapFocus={swapFocus}
+          amount={
+            inputType === "fiat"
+              ? type === "market" && tab === "sell"
+                ? trimPlaceholderZeros(
+                    (expectedOutput ?? new Dec(0)).toString()
+                  )
+                : fiatAmount
+              : type === "market" && tab === "buy"
+              ? trimPlaceholderZeros((expectedOutput ?? new Dec(0)).toString())
+              : tokenAmount
+          }
+          setter={(v) =>
+            inputType === "fiat"
+              ? setAmountSafe("fiat", v)
+              : setAmountSafe("token", v)
+          }
+          disableSwitching={disableSwitching}
+          loading={
+            inputType === "fiat"
+              ? tab === "sell" && type === "market" && expectedOutputLoading
+              : tab === "buy" && type === "market" && expectedOutputLoading
+          }
+        />
+      ))}
+      <button
+        className="absolute right-4 top-3 flex items-center justify-center rounded-5xl border border-osmoverse-700 bg-osmoverse-1000 py-1.5 px-3 opacity-50 transition-opacity hover:opacity-100"
+        onClick={toggleMax}
+      >
+        <span
+          className={classNames("body2 text-wosmongton-200", {
+            "text-rust-300": insufficientFunds,
+          })}
+        >
+          Max
+        </span>
+      </button>
+    </div>
+  );
+};
+
+type AutoInputProps = {
+  focused: FocusedInput;
+  swapFocus: () => void;
+  setter: (v: string) => void;
+  amount: string;
+  type: "fiat" | "token";
+  loading: boolean;
+} & Omit<
+  LimitInputProps,
+  | "onChange"
+  | "price"
+  | "tokenAmount"
+  | "setMarketAmount"
+  | "quoteAssetPrice"
+  | "expectedOutput"
+  | "expectedOutputLoading"
+>;
+
+function AutoInput({
+  focused,
+  baseAsset,
+  swapFocus,
+  amount,
+  setter,
+  type,
+  disableSwitching,
+  loading,
+}: AutoInputProps) {
+  const { isMobile } = useWindowSize();
+  const currentTypeEnum = useMemo(
+    () => (type === "fiat" ? FocusedInput.FIAT : FocusedInput.TOKEN),
+    [type]
+  );
+
+  const isFocused = useMemo(
+    () => focused === currentTypeEnum,
+    [currentTypeEnum, focused]
+  );
+
+  const oppositeTypeEnum = useMemo(
+    () => (type === "fiat" ? FocusedInput.TOKEN : FocusedInput.FIAT),
+    [type]
+  );
+
+  const scale = useMemo(
```

Comment:
Similar to boolean, string values don't need to be memoed since there's no references to maintain. Only memo if `calcScale` is expensive.